tools/utils.hh

```diff
@@ -61,4 +62,175 @@ const Op& get_selected_operation(int& ac, char**& av, const std::vector<Op>& ope
+        const std::vector<operation_option>* operation_options = nullptr;
+        const std::vector<app_template::positional_option>* operation_positional_options = nullptr;
+    };
+
+private:
+    config _cfg;
+
+public:
+    tool_app_template(config cfg)
+        : _cfg(std::move(cfg))
+    { }
+
+    int run(int argc, char** argv, noncopyable_function<int(const operation&, const boost::program_options::variables_map&)> main_func) {
+        const operation* found_op = nullptr;
+        if (std::strcmp(argv[1], "--help") != 0 && std::strcmp(argv[1], "-h") != 0) {
+            found_op = &tools::utils::get_selected_operation(argc, argv, _cfg.operations, "operation");
+        }
+
+        app_template::seastar_options app_cfg;
+        app_cfg.name = _cfg.name;
+
+        if (found_op) {
+            app_cfg.description = format("{}\n\n{}\n", found_op->summary(), found_op->description());
+        } else {
+            app_cfg.description = _cfg.description;
+        }
+
+        tools::utils::configure_tool_mode(app_cfg, _cfg.logger_name);
+
+        app_template app(std::move(app_cfg));
+
+        if (_cfg.global_options) {
+            auto& global_desc = app.get_options_description();
+            for (const auto& go : *_cfg.global_options) {
+                go.add_option(global_desc);
+            }
+        }
+        if (_cfg.global_positional_options) {
+            for (const auto& gpo : *_cfg.global_positional_options) {
+                app.add_positional_options({gpo});
+            }
+        }
+
+        if (found_op) {
+            boost::program_options::options_description op_desc(found_op->name());
+            size_t options_added = 0;
+            for (const auto& opt_name : found_op->available_options()) {
+                if (_cfg.operation_options) {
+                    auto it = std::ranges::find_if(*_cfg.operation_options, [&] (const operation_option& opt) { return opt.name() == opt_name; });
+                    if (it != _cfg.operation_options->end()) {
+                        it->add_option(op_desc);
+                        ++options_added;
+                        continue;
+                    }
+                }
+                if (_cfg.operation_positional_options) {
+                    auto it = std::ranges::find_if(*_cfg.operation_positional_options, [&] (const app_template::positional_option& opt) { return opt.name == opt_name; });
+                    if (it != _cfg.operation_positional_options->end()) {
+                        app.add_positional_options({*it});
+                        continue;
+                    }
+                }
+                std::abort(); // failed to look-up option
+            }
+            if (options_added) {
+                app.get_options_description().add(op_desc);
+            }
+        }
+
+        return app.run(argc, argv, [this, &main_func, &app, found_op] {
+            return async([this, &main_func, &app, found_op] {
```

Comment:
Why not consider the main_func() as a future<>-returning one from the very beginning and avoid using `async()`? So that the tool code is hopefully encouraged to make full use of coroutines

Reply:
`async()` is still the most convenient way to do async coding:
* you can wait for future anywhere: constructors, destructors, operators, catch clauses
* RAII works the same as in sync code

Therefore, if performance is not a concern, I still always prefer `async()` over coroutines.

Reply:
Probably threads are a little faster than coroutines, especially when nesting calls. The primary advantage of coroutines is reduced memory usage.

I still like to return a future in the interface and explicitly call seastar::async(), but it's not really necessary.